### DISCUSSION TITLE
CQRS with Rest API

### DISCUSSION THREAD

**Discussion by NghiaNguyen170192** [jasontaylordev/CleanArchitecture#457] [Q&A] [answered]
I'm learning Clean Architecture and what I found confusing is the CQRS Update Command with Rest API. 
When doing the Update on Rest API, Id is in the header and the update request is on the Body Request. 

I check the `TodoListsController` and I see that the id is in 2 places, in the request header and in `UpdateTodoListCommand`.

So how can I have an update request using `UpdateTodoListCommand` without containing property `Id` in the Command and just have Id in the request header?

**Comment by andrewtsw** [comment, marked as answer]
Hi @NghiaNguyen170192 
I believe the best way to fix it - do not use commands as a DTO.
Co, create `UpdateTodoListDto` without `id` and use it as a parameter of the controller method. 
Then create a command and put `id` and DTO to it. And a command  send to handler.
So no `id` duplication.
What do you think about this approach?

**Reply by NghiaNguyen170192** [reply]
> Hi @NghiaNguyen170192 I believe the best way to fix it - do not use commands as a DTO. Co, create `UpdateTodoListDto` without `id` and use it as a parameter of the controller method. Then create a command and put `id` and DTO to it. And a command send to handler. So no `id` duplication. What do you think about this approach?

> As @andrewtsw mentioned, it is the same approach which @vkhorikov wrote in his article https://enterprisecraftsmanship.com/posts/cqrs-commands-part-domain-model/ Use dtos in controller actions to map to commands, instead of using direct commands.

I dont think this is the approach to go. If you have a large number of entities, then you have twice number of models, 1 for DTO and 1 for command. And then you have to manage 2 models at the same time if requirement is changed. 
Also , it breaks the convention using CQRS in the `ApiController`
If so, do I need to have a complete bunch of DTOs for CRUD that match with Commands?

Another question is that the Rest API also has partial update on an entity. So do I have to create each command corresponding for each partial update request?

**Reply by neman** [reply]
It depends of context.
I have production app with this template using commands in controller.

But you can decouple command from api contract. So you can use the same command for update, but only expose endpoints with different dtos to the same command (if it is generic update command and not specific e.g. UpdateCustomer vs ChangeCustomerAddress)

As always, there is no silver bullet and it depends. Yes, you will have multiple models, dtos and commands, but you can use automapper to map. After all command is just a dto (and do not forget not to map command to domain, it is no no if you are following DDD approach, but that is another theme)

**Comment by neman** [comment]
As @andrewtsw mentioned, it is the same approach which @vkhorikov wrote in his article https://enterprisecraftsmanship.com/posts/cqrs-commands-part-domain-model/
Use dtos in controller actions to map to commands, instead of using direct commands.
